### PR TITLE
Fix issue #100 and #99

### LINKED ISSUE #99: Errors after upgrade
Hello,
today I've upgraded jwt-cpp and I have compiler errors:

```
[ 94%] Building CXX object CMakeFiles/nms.dir/src/controller/UserController.cpp.o
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/service/tokenService.hpp:7,
                 from /home/marcin/projects/nms/back/src/service/tokenService.cpp:5:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In constructor ‘jwt::algorithm::ed448::ed448(const string&, const string&, const string&, const string&)’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:1316:81: error: ‘EVP_shake256’ was not declared in this scope; did you mean ‘EVP_sha256’?
 1316 |     : eddsa(public_key, private_key, public_key_password, private_key_password, EVP_shake256, "EdDSA")
      |                                                                                 ^~~~~~~~~~~~
      |                                                                                 EVP_sha256
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/service/tokenService.hpp:7,
                 from /home/marcin/projects/nms/back/src/service/userCache.hpp:9,
                 from /home/marcin/projects/nms/back/src/service/userCache.cpp:2:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In constructor ‘jwt::algorithm::ed448::ed448(const string&, const string&, const string&, const string&)’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:1316:81: error: ‘EVP_shake256’ was not declared in this scope; did you mean ‘EVP_sha256’?
 1316 |     : eddsa(public_key, private_key, public_key_password, private_key_password, EVP_shake256, "EdDSA")
      |                                                                                 ^~~~~~~~~~~~
      |                                                                                 EVP_sha256
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/service/tokenService.hpp:7,
                 from /home/marcin/projects/nms/back/src/service/userCache.hpp:9,
                 from /home/marcin/projects/nms/back/src/controller/UserController.hpp:7,
                 from /home/marcin/projects/nms/back/src/controller/UserController.cpp:1:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In constructor ‘jwt::algorithm::ed448::ed448(const string&, const string&, const string&, const string&)’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:1316:81: error: ‘EVP_shake256’ was not declared in this scope; did you mean ‘EVP_sha256’?
 1316 |     : eddsa(public_key, private_key, public_key_password, private_key_password, EVP_shake256, "EdDSA")
      |                                                                                 ^~~~~~~~~~~~
      |                                                                                 EVP_sha256
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/dto/apiDTO/AuthResponse.hpp:5,
                 from /home/marcin/projects/nms/back/src/dto/apiDTO/AuthResponse.cpp:2:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In constructor ‘jwt::algorithm::ed448::ed448(const string&, const string&, const string&, const string&)’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:1316:81: error: ‘EVP_shake256’ was not declared in this scope; did you mean ‘EVP_sha256’?
 1316 |     : eddsa(public_key, private_key, public_key_password, private_key_password, EVP_shake256, "EdDSA")
      |                                                                                 ^~~~~~~~~~~~
      |                                                                                 EVP_sha256
make[3]: *** [CMakeFiles/nms.dir/build.make:135: CMakeFiles/nms.dir/src/service/tokenService.cpp.o] Błąd 1
make[3]: *** Oczekiwanie na niezakończone zadania....
make[3]: *** [CMakeFiles/nms.dir/build.make:148: CMakeFiles/nms.dir/src/service/userCache.cpp.o] Błąd 1
make[3]: *** [CMakeFiles/nms.dir/build.make:96: CMakeFiles/nms.dir/src/dto/apiDTO/AuthResponse.cpp.o] Błąd 1
In file included from /home/marcin/projects/nms/back/src/dto/Token.hpp:5,
                 from /home/marcin/projects/nms/back/src/service/tokenService.hpp:7,
                 from /home/marcin/projects/nms/back/src/service/userCache.hpp:9,
                 from /home/marcin/projects/nms/back/src/controller/InitController.hpp:12,
                 from /home/marcin/projects/nms/back/src/controller/InitController.cpp:1:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘std::string jwt::algorithm::eddsa::sign(const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:970:9: error: ‘EVP_DigestSign’ was not declared in this scope; did you mean ‘EVP_DigestSignInit’?
  970 |     if (EVP_DigestSign(ctx.get(),
      |         ^~~~~~~~~~~~~~
      |         EVP_DigestSignInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In member function ‘void jwt::algorithm::eddsa::verify(const string&, const string&, std::error_code&) const’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:998:16: error: ‘EVP_DigestVerify’ was not declared in this scope; did you mean ‘EVP_DigestVerifyInit’?
  998 |     auto res = EVP_DigestVerify(ctx.get(),
      |                ^~~~~~~~~~~~~~~~
      |                EVP_DigestVerifyInit
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h: In constructor ‘jwt::algorithm::ed448::ed448(const string&, const string&, const string&, const string&)’:
/home/marcin/projects/nms/back/lib/jwt/include/jwt-cpp/jwt.h:1316:81: error: ‘EVP_shake256’ was not declared in this scope; did you mean ‘EVP_sha256’?
 1316 |     : eddsa(public_key, private_key, public_key_password, private_key_password, EVP_shake256, "EdDSA")
      |                                                                                 ^~~~~~~~~~~~
      |                                                                                 EVP_sha256
```

What's wrong?

### LINKED ISSUE #100: LibreSSL does not support EVP_DigestSign and EVP_DigestVerify
The new EdDSA support uses calls to EVP_DigestSign/EVP_DigestVerify which are not supported by LibreSSL.
We dont officialy support LibreSSL but since it seems to otherwise compile fine and its an pretty easy fix I think we should remove calls to those convenience functions and use EVP_DigestSignInit and EVP_DigestSignFinal instead.
We might also consider adding a ci test to make sure everything actually works with libressl as it should.

### PR DESCRIPTION
**Pull request by Thalhammer**
closes #99 
closes #100

### CODE REVIEW

File: include/jwt-cpp/jwt.h [hunk 1]
```diff
@@ -967,12 +967,28 @@ namespace jwt {
 				size_t len = EVP_PKEY_size(pkey.get());
 				std::string res(len, '\0');
 
+				// LibreSSL is the special kid in the block, as it does not support EVP_DigestSign.
+				// OpenSSL on the otherhand does not support using EVP_DigestSignUpdate for eddsa, which is why we end up with this mess.
+				#ifdef LIBRESSL_VERSION_NUMBER
+				ERR_clear_error();
+				auto e = EVP_DigestSignUpdate(ctx.get(), reinterpret_cast<const unsigned char*>(data.data()), data.size());
+				if(e != 1) {
```
Comment (prince-chrismc) [inline include/jwt-cpp/jwt.h:975]: To keep it consistent

```suggestion
				if(EVP_DigestSignUpdate(ctx.get(), reinterpret_cast<const unsigned char*>(data.data()), data.size()) != 1) {
```

File: include/jwt-cpp/jwt.h [hunk 2]
```diff
@@ -967,12 +967,28 @@ namespace jwt {
 				size_t len = EVP_PKEY_size(pkey.get());
 				std::string res(len, '\0');
 
+				// LibreSSL is the special kid in the block, as it does not support EVP_DigestSign.
+				// OpenSSL on the otherhand does not support using EVP_DigestSignUpdate for eddsa, which is why we end up with this mess.
```
Comment (prince-chrismc) [inline include/jwt-cpp/jwt.h:971]: 🤣 Love it!

Only because I usually think about this at work, but should we be a bit more formal? 🤔

**Review by prince-chrismc — COMMENTED** [review]
Does this mean we aim to official support libresll? or is the intention to leave it as unofficially compatible.

I assume there was the only thing preventing it compiling.

File: include/jwt-cpp/jwt.h (hunk 2 above)
Comment (Thalhammer) [reply include/jwt-cpp/jwt.h:971]: Nope, it's a fact so they may as well know what I think about it.

**Comment by Thalhammer** [comment]
> Does this mean we aim to official support libresll? or is the intention to leave it as unofficially compatible.
> I assume there was the only thing preventing it compiling.

Seems like it compiles fine now, however I don't have any good way to test as all my systems use openssl.
Since we will end up having to support it anyway, we might as well officially support it once we have a proper CI setup that runs the tests and build against libressl.

**Pull request merged**